clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java

```diff
@@ -0,0 +1,238 @@
+                                         final Optional<String> topic,
+                                         final long retryBackoffMs,
+                                         final long retryBackoffMaxMs) {
+            super(logContext, TopicMetadataRequestState.class.getSimpleName(), retryBackoffMs,
+                retryBackoffMaxMs);
+            this.future = new CompletableFuture<>();
+            this.topic = topic;
+        }
+
+        /**
+         * prepare the metadata request and return an
+         * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.UnsentRequest} if needed.
+         */
+        private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTimeMs) {
+            if (!this.canSendRequest(currentTimeMs)) {
+                return Optional.empty();
+            }
+            this.onSendAttempt(currentTimeMs);
+
+            final MetadataRequest.Builder request;
+            request = topic.map(t -> new MetadataRequest.Builder(Collections.singletonList(t), allowAutoTopicCreation))
+                    .orElseGet(MetadataRequest.Builder::allTopics);
+
+            final NetworkClientDelegate.UnsentRequest unsent = createUnsentRequest(request);
+            return Optional.of(unsent);
+        }
+
+        private NetworkClientDelegate.UnsentRequest createUnsentRequest(
+                final MetadataRequest.Builder request) {
+            return new NetworkClientDelegate.UnsentRequest(
+                    request,
+                    Optional.empty(),
+                    this::processResponseOrException
+            );
+        }
+
+        private void processResponseOrException(final ClientResponse response,
+                                                final Throwable exception) {
+            long responseTimeMs = System.currentTimeMillis();
+            if (exception != null) {
+                handleException(exception, responseTimeMs);
+                return;
+            }
+            handleResponse(response, responseTimeMs);
+        }
+
+        private void handleException(final Throwable exception, final long responseTimeMs) {
+            if (exception instanceof TimeoutException || !(exception instanceof RetriableException)) {
+                completeFutureAndRemoveRequest(new KafkaException(exception));
+            } else {
+                // We continue to retry on RetriableException
+                onFailedAttempt(responseTimeMs);
+            }
+        }
+
+        private void handleResponse(final ClientResponse response, final long responseTimeMs) {
+            try {
+                Map<String, List<PartitionInfo>> res = handleTopicMetadataResponse((MetadataResponse) response.responseBody());
+                future.complete(res);
+                inflightRequests.remove(topic);
+            } catch (RetriableException e) {
+                onFailedAttempt(responseTimeMs);
```

Comment:
The case that I am wondering is the following.

DefaultBackgroundThread is in a retry loop for an inflightRequest. PrototypeAsyncConsumer.partitionsFor eventually calls CompletableApplicationEvent.get(Timer timer) and times out. Do we still keep the request in inflightRequests at that point? Is that a concern?

Reply:
Thanks Jun for raising this point - I think what we should do is to complete the request with a non-retriable exception to prevent the request from retrying. Otherwise - the request manager will never remove the request from the inflightRequest.